internal/server/v1/observations/derived_series.go

```diff
@@ -0,0 +1,507 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package observations implements for /v1/derived/observations/series/...
+package observations
+
+import (
+	"context"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"os"
+	"reflect"
+	"sort"
+	"strings"
+
+	pb "github.com/datacommonsorg/mixer/internal/proto"
+	"github.com/datacommonsorg/mixer/internal/server/stat"
+	"github.com/datacommonsorg/mixer/internal/store"
+)
+
+// DerivedSeries implements API for Mixer.DerivedObservationsSeries.
+func DerivedSeries(
+	ctx context.Context,
+	in *pb.DerivedObservationsSeriesRequest,
+	store *store.Store,
+) (*pb.DerivedObservationsSeriesResponse, error) {
+	resp := &pb.DerivedObservationsSeriesResponse{}
+	entity := in.GetEntity()
+
+	// Parse the formula to extract all the variables, used for reading data from BT.
+	varCalculator, err := newVarCalculator(in.GetFormula())
+	if err != nil {
+		return resp, err
+	}
+	statVars := varCalculator.statVars()
+
+	// Read data from BT.
+	btData, err := stat.ReadStatsPb(
+		ctx, store.BtGroup, []string{entity}, statVars)
+	if err != nil {
+		return resp, err
+	}
+	entityData, ok := btData[entity]
+	if !ok {
+		return resp, err
+	}
+
+	// Calculate.
+	result, err := varCalculator.calculate(entityData)
+	if err != nil {
+		return resp, err
+	}
+	for _, p := range result.points {
+		resp.Observations = append(resp.Observations, &pb.PointStat{
+			Date:  p.GetDate(),
+			Value: p.GetValue(),
+		})
+	}
+
+	return resp, nil
+}
+
+type varCalculatorSeries struct {
```

Comment:
a general comment, can remove "var" prefix in most of the struct, function names here? On one hand, it's not really stat var calculator, but observation calculator; also calculator is pretty unique in mixer context.

Reply:
You can just use pb.SourceSeries for this struct, and the ranking is free from https://github.com/datacommonsorg/mixer/blob/master/internal/server/ranking/ranking.go#L254

Reply:
done - removed 'var' from name.

As discussed, will hold off for using pb.SourceSeries for now, and consider modifying pb.SourceSeries to use repeated/sorted field instead of map for 'val'.